### PR TITLE
CIRC-1343 Items in transit report performance: fetch service points

### PR DESCRIPTION
**Pull request by roman-barannyk**
## Purpose
Part of the new items-in-transit report implementation.

## Approach
- Collect service points ids from items, loans, requests;
- Use the same fetcher as the original implementation to fetch service points in batches;

Resolves: [CIRC-1343](https://issues.folio.org/browse/CIRC-1343)

### CODE REVIEW

File: src/main/java/org/folio/circulation/infrastructure/storage/loans/LoanRepository.java [hunk 1]
```diff
@@ -224,6 +225,16 @@ else if(loans.getTotalRecords() == 1) {
       .thenComposeAsync(loans -> itemRepository.fetchItemsFor(loans, Loan::withItem));
   }
 
+  public CompletableFuture<Result<MultipleRecords<Loan>>> findByItemIds(
+    Collection<String> itemIds) {
+
+    Result<CqlQuery> statusQuery = exactMatch("itemStatus", IN_TRANSIT.getValue());
```
Comment (sherzodn) [inline src/main/java/org/folio/circulation/infrastructure/storage/loans/LoanRepository.java:231]: Use already defined constant ITEM_STATUS

Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/infrastructure/storage/loans/LoanRepository.java:231]: Have you merged the PR?

Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/infrastructure/storage/loans/LoanRepository.java]: Found, updated, thanks

File: src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java [hunk 2]
```diff
@@ -2,25 +2,38 @@
...
 import static org.folio.circulation.support.results.Result.succeeded;
+import static org.folio.circulation.support.results.ResultBinding.mapResult;
 
+import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
```
Comment (sherzodn) [inline src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java:17]: Remove this duplicated import.

File: src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java [hunk 3]
```diff
@@ -2,25 +2,38 @@
...
 import static java.util.concurrent.CompletableFuture.completedFuture;
 import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toSet;
 import static org.folio.circulation.domain.ItemStatus.IN_TRANSIT;
 import static org.folio.circulation.support.results.Result.combineAll;
 import static org.folio.circulation.support.results.Result.succeeded;
+import static org.folio.circulation.support.results.ResultBinding.mapResult;
 
+import java.util.ArrayList;
+import java.util.Collection;
```
Comment (sherzodn) [inline src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java:12]: Remove this duplicated import

File: src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java (hunk 2 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java:16]: done

File: src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java (hunk 3 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java:12]: done

File: src/test/java/org/folio/circulation/services/ItemsInTransitReportServiceTest.java [hunk 4]
```diff
@@ -1,22 +1,29 @@
...
 import org.folio.circulation.domain.Item;
 import org.folio.circulation.domain.ItemsReportFetcher;
+import org.folio.circulation.domain.Loan;
 import org.folio.circulation.domain.Location;
 import org.folio.circulation.domain.MultipleRecords;
+import org.folio.circulation.infrastructure.storage.ServicePointRepository;
 import org.folio.circulation.infrastructure.storage.inventory.ItemReportRepository;
 import org.folio.circulation.infrastructure.storage.inventory.ItemRepository;
+import org.folio.circulation.infrastructure.storage.loans.LoanRepository;
+import org.folio.circulation.infrastructure.storage.inventory.ItemRepository;
```
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/test/java/org/folio/circulation/services/ItemsInTransitReportServiceTest.java]: updated

**Review by sherzodn — APPROVED** [review]

File: src/main/java/org/folio/circulation/infrastructure/storage/ServicePointRepository.java [hunk 5]
```diff
@@ -177,6 +177,12 @@ public ServicePointRepository(Clients clients) {
           }));
   }
 
+  public CompletableFuture<Result<MultipleRecords<ServicePoint>>> findServicePointsByIds(
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/infrastructure/storage/ServicePointRepository.java:180]: collection

File: src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java [hunk 6]
```diff
@@ -149,11 +149,46 @@
...
-  // Needs to fetch all service points for items, loans and requests
   private CompletableFuture<Result<ItemsInTransitReportContext>> fetchServicePoints(
-    Result<ItemsInTransitReportContext> context) {
+    ItemsInTransitReportContext context) {
 
-    return completedFuture(context);
+    Collection<Item> items = context.getItems().values();
+    Stream<String> itemInTransitDestinationServicePointIds = items.stream()
+      .map(Item::getInTransitDestinationServicePointId)
+      .filter(Objects::nonNull);
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java:158]: remove

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**